model/task_history.go

```diff
@@ -394,17 +389,34 @@ func (self *taskHistoryIterator) GetDistinctTestNames(numCommits int) ([]string,
 			{"$unwind": fmt.Sprintf("$%v", testResultsKey)},
 			{"$group": bson.M{"_id": fmt.Sprintf("$%v.%v", testResultsKey, task.TestResultTestFileKey)}},
 		},
+		opts,
 	)
-
-	var output []bson.M
-
-	if err = pipeline.All(&output); err != nil {
-		return nil, errors.WithStack(err)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting cursor")
+	}
+	if cursor == nil {
+		return nil, errors.New("nil cursor returned")
 	}
 
-	names := make([]string, 0)
-	for _, doc := range output {
-		names = append(names, doc["_id"].(string))
+	names := []string{}
+	for cursor.Next(ctx) {
+		select {
+		case <-ctx.Done():
+			_ = cursor.Close(ctx)
+			return nil, errors.New("context cancelled")
+		default:
+			res := bson.M{}
+			if err := cursor.Decode(&res); err != nil {
+				return nil, errors.Wrapf(err, "error decoding result")
```

Comment:
it looks like in this case we won't close the cursor.
Maybe we should defer closing the cursor right after we get it? [The docs](https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#Cursor.Close) claim the Close is idempotent.

Reply:
you can probably defer closing it for simplicity

Reply:
Good idea!